src/plot.cpp

```diff
@@ -1581,6 +1532,303 @@ void ProjectionPlot::set_field_of_view(pugi::xml_node node)
+        return;
+      }
+
+      // Advance particle, prepare for next intersection
+      for (int lev = 0; lev < n_coord(); ++lev) {
+        coord(lev).r += dist_.distance * coord(lev).u;
+      }
+      surface() = dist_.surface_index;
+      n_coord_last() = n_coord();
+      n_coord() = dist_.coord_level;
+      if (dist_.lattice_translation[0] != 0 ||
+          dist_.lattice_translation[1] != 0 ||
+          dist_.lattice_translation[2] != 0) {
+        cross_lattice(*this, dist_, settings::verbosity >= 10);
+      }
+
+    } else {
+      first_surface_ = RayTracePlot::advance_to_boundary_from_void(*this);
+      intersection_found_ = first_surface_ != -1; // -1 if no surface found
+    }
+    event_counter_++;
+    if (event_counter_ > MAX_INTERSECTIONS)
+      fatal_error("Infinite loop in ray traced plot");
+  }
+}
+
+void ProjectionRay::on_intersection()
+{
+  // This allows drawing wireframes with surface intersection
+  // edges on the model boundary for the same cell.
+  if (first_inside_model()) {
+    line_segments_.emplace_back(
+      plot_.color_by_ == PlottableInterface::PlotColorBy::mats
+        ? material()
+        : lowest_coord().cell,
+      0.0, first_surface());
+  }
+
+  // This records a tuple with the following info
+  //
+  // 1) ID (material or cell depending on color_by_)
+  // 2) Distance traveled by the ray through that ID
+  // 3) Index of the intersected surface (starting from 1)
+  line_segments_.emplace_back(
+    plot_.color_by_ == PlottableInterface::PlotColorBy::mats
+      ? material()
+      : lowest_coord().cell,
+    dist().distance, std::abs(dist().surface_index));
+}
+
+void PhongRay::on_intersection()
+{
+  // Check if we hit an opaque material or cell
+  int hit_id = plot_.color_by_ == PlottableInterface::PlotColorBy::mats
+                 ? material()
+                 : lowest_coord().cell;
+
+  // If we are reflected and have advanced beyond the camera,
+  // the ray is done. This is checked here because we should
+  // kill the ray even if the material is not opaque.
+  if (reflected_ && (r() - plot_.camera_position()).dot(u()) >= 0.0) {
+    stop();
+    return;
+  }
+
+  // Anything that's not opaque has zero impact on the plot.
+  if (!plot_.is_id_opaque(hit_id))
+    return;
+
+  if (!reflected_) {
+
+    // reflect the particle and set the color to be colored by
+    // the normal or the diffuse lighting contribution
+    reflected_ = true;
+    result_color_ = plot_.colors_[hit_id];
+    Direction to_light = plot_.light_location_ - r();
+    to_light /= to_light.norm();
+
+    // Not sure what can cause this error. Color as an error and proceed
+    // from there. It seems to happen only for a few pixels on the outer
+    // boundary of a hex lattice. TODO
+    //
+    // We cannot detect it in the outer loop, and it only matters here, so
+    // that's why the error handling is a little different than for a lost
+    // ray.
+    if (i_surface() == -1) {
```

Comment:
This might be another issue that could be solved by using our particle transport mechanisms for surface intersection and crossing.